### PR TITLE
Nigel options consolidation

### PR DESCRIPTION
**Pull request by nigel-dick**
abstracts code for connection objects to remove redundancies

### CODE REVIEW

File: data_io/data_io.py [hunk 1]
```diff
@@ -482,46 +537,27 @@ def write(
...
-
-      write_options[option] = value
-
-    # Set format and mode
-    writer = (
-      writer
-      .format(write_options['format'])
-      .mode(write_options['mode'])
+    
+    writer = super().set_writer(
```
Comment (chris-bishop-tfs) [inline data_io/data_io.py:541]: Why `super()` instead of `self`?

File: data_io/data_io.py [hunk 2]
```diff
@@ -385,14 +384,91 @@ def __init__(self, *largs, **kwargs):
...
+    return options
+  
+  def set_options(self, r_w, options):
+    # Set options for the reader object
+    for option, value in options.items():
+      r_w = r_w.option(option, value)
+      
+    return r_w
+  
+  def set_reader(
```
Comment (chris-bishop-tfs) [inline data_io/data_io.py:420]: `set` and `get` have special meanings in python land - let's reconsider the verb we use. Maybe `build_reader`? Not sure.

File: data_io/data_io.py [hunk 3]
```diff
@@ -385,14 +384,91 @@ def __init__(self, *largs, **kwargs):
...
+
+    reader = spark.read.format(read_options['format'])
+    reader = self.set_options(reader, read_options)
+
+    return reader
+    
+  def set_writer(
+    self,
+    data,
+    default_write_options = None,
```
Comment (chris-bishop-tfs) [inline data_io/data_io.py:440]: See PEP8 guidelines for parameter spacing

https://peps.python.org/pep-0008/#function-and-variable-names

image

**Review by chris-bishop-tfs — CHANGES_REQUESTED** [review]
Great work, @nigel-dick. Lots of redundancy removed.

Comments are mostly in the realm of specifics/nitpicking.

Also, can you add in variable annotations to the methods? Ping me if this is unclear.

File: data_io/data_io.py (hunk 2 above)
Comment (nigel-dick) [reply data_io/data_io.py]: methods renamed

**Comment by nigel-dick** [comment]
variable annotations added

File: data_io/data_io.py (hunk 1 above)
Comment (nigel-dick) [reply data_io/data_io.py]: At some point while making these updates, using `self` was throwing an error. But, I replaced all the `super().` calls with `self.` and there are no errors. Must have been related to some other unknown bug I introduced that was fixed along the way.

File: data_io/data_io.py [hunk 4]
```diff
@@ -385,20 +385,97 @@ def __init__(self, *largs, **kwargs):
...
+    for option, value in kwargs.items():
+      options[option] = value
+
+    # This is clunky, but we needed a way to support query/dbtable
+    if not(isinstance(self, S3Connection)) and is_read and ('query' not in options.keys() and 'dbtable' not in options.keys()):
+      options['query'] = f'SELECT * FROM {self.location.schema}.{self.location.table}'
+    
+    return options
+  
+  def implement_options(self, r_w: Union[DataFrameWriter, DataFrameReader], options: dict) -> Union[DataFrameWriter, DataFrameReader]:
```
Comment (chris-bishop-tfs) [inline data_io/data_io.py:414]: Doc strings

**Review by chris-bishop-tfs — APPROVED** [review]

**Pull request merged**